massenergize/api
# Supporting Feature Flags

This PR introduces Feature Flags support for Massenergize according to the spec here:
https://paper.dropbox.com/doc/Feature-Flags--BkpUY1IpM~wj4Dr1798_Ku2RAQ-E3KtCfp5ZQN2fgzTjB1DJ
The PR works with [Frontend Admin PR](https://github.com/massenergize/frontend-admin/pull/637)

#### Details 
With this PR, you can now fetch all feature flags turned on for 
1. An admin 
2. A  normal user

#### There are now new routes to
1. Add feature flag 
2. Delete a flag 
3. Just list all the available flags for super-admin display 
4. [Slight changes to Sam's `get_feature_flags`.] Now that route will load all features available to a user, and all the communities they have anything to do with, depending on whether they are on the admin site, or user portal. (Check comments in store/get_feature_flags for breakdown)


#### Testing Steps 
1. The Repo is still in local mode, but check to make sure, then run `... makemigrations` and `migrate` for the new db changes
6. For manual testing, Go to django admin and create a feature flag or step into the [sadmin-feature-flags-implementation ](https://github.com/massenergize/frontend-admin/pull/637) branch on the admin repo to test.


### Up next 
* Test files for new routes I have added ( all features listed above have been manually tested though)


### Note 
* Specific communities, and users that are selected when creating a feature maybe used as "whitelist" or "blacklist" depending on whether the "audience" is set to `"SPECIFIC or ALL_EXCEPT"`\
* Also, the relationship between communities and users on a feature is currently implemented with OR logic. (cant say there is any special reason 💀 🤣) However, if we find the downsides, or see more advantages with AND logic, it can easily be changed.

Feedback wld be great 🤝
